### PR TITLE
[HTTPCLIENT-2042] AuthScope uses uppercase for scheme name whereas th…

### PR DESCRIPTION
**Pull request by michael-o**
…e rest of the code uses the lowercase version

The auth scheme name is now consistently processed either in lowercse or
with String#equalsIgnoreCase().

@garydgregory @ok2c Please share your opinion. Another approach would also be to leave as-is because `AuthScheme#getName()` is retained as-is and lowercased where ever appropriate. That would be even better for consistency. Shall I rework the issue and PR then?

### CODE REVIEW

**Comment by ok2c** [comment]
> Another approach would also be to leave as-is because AuthScheme#getName() is retained as-is and lowercased where ever appropriate. That would be even better for consistency.

@michael-o I am not sure I understand the need to make any changes of what so ever, so leaving things as is would be my preference.

**Comment by michael-o** [comment]
> 
> 
> > Another approach would also be to leave as-is because AuthScheme#getName() is retained as-is and lowercased where ever appropriate. That would be even better for consistency.
> 
> @michael-o I am not sure I understand the need to make any changes of what so ever, so leaving things as is would be my preference.

Let me rephrase the confusing inconsistency I see: `AuthScheme#getName()` returns the scheme name as-is while `AuthScope#getScheme()` returns it in uppercase for no apparent reason.
It that better to understand?

**Comment by ok2c** [comment]
@michael-o I still have no idea what you see as inconsistency here. Auth scheme names are case-insensitive much so as domain names. Please go ahead and change whatever feel like changing as long as it gets done fast.

**Comment by garydgregory** [comment]
Should we not keep the auth scheme strings to the two styles we already have? IOW:
- The upper case names as the enum names in org.apache.hc.client5.http.auth.AuthSchemes: BASIC, DIGEST, and so on.
- The mixed case names as the 'ident' field of org.apache.hc.client5.http.auth.AuthSchemes: "Basic", "Digest", and so on.

Question... I see we have these 2 ctors:
- org.apache.hc.client5.http.auth.AuthScope.AuthScope(HttpHost, String, String)
- org.apache.hc.client5.http.auth.AuthScope.AuthScope(String, String, int, String, String)
These both take a schemeName. 
Why not avoid confusion and type this as an AuthScheme? Or both?
This goes back to the email thread about how extenstible AuthScheme(s) are suppose to be. 
This needs clarification before we continue IMO.

**Comment by michael-o** [comment]
@garydgregory All of your analysis leads me to the conclusion that the enum *and* strings do not make sense together. Always consider that enums are designed to be exhaustive. Our support is not exhaustive... I will work on a proposal.

**Comment by ok2c** [comment]
> Always consider that enums are designed to be exhaustive

@michael-o Out of curiosity, who decreed that?

**Comment by michael-o** [comment]
> 
> 
> > Always consider that enums are designed to be exhaustive
> 
> @michael-o Out of curiosity, who decreed that?

@ok2c, in my opinion this is a convention. Since a client cannot extend or create custom enum instances, the list remains unmodifiable, thus always complete. Therefore, you can use `==` for comparsion because one enum value exists in the class loader.

I cannot say

```
public enum EnterpriseAuthSchemes extends AuthSchemes {

  COOL("Cool")

}
```

**Comment by ok2c** [comment]
@michael-o But enums can be extended in future releases without breaking binary compatibility. They can never be assumed to be complete and final. 

This whole enum renaming stuff is nothing more than assumptions and personal fancy.

**Comment by garydgregory** [comment]
We are getting a bit side tracked here but an enum is finite by definition. Arguing they can be extended by a future release is not a valid point IMO. I can't write code today (a switch for example) for an enum value that will only exist in the future. This is obviously completely different than using a String where you can write anything you want.

**Comment by michael-o** [comment]
> 
> 
> @michael-o But enums can be extended in future releases without breaking binary compatibility. They can never be assumed to be complete and final.

Given with the current enum semantics. They are final implicitly. I cannot see in the future whether this will change. I remember my programming lecture when Sun introduced enums back then and the professor said they did not make it extendable for a good reason.

Making assumptions about likely possible features will make any design decision really hard.

> This whole enum renaming stuff is nothing more than assumptions and personal fancy.

Partially right, but can't consistency be objective in this case?

> We are getting a bit side tracked here but an enum is finite by definition. Arguing they can be extended by a future release is not a valid point IMO. I can't write code today (a switch for example) for an enum value that will only exist in the future. This is obviously completely different than using a String where you can write anything you want.

+1, totally agree. Future extensions will break code. This is actually something Sun never wanted. Therefore generics have been implemented in a reified way.

**Comment by garydgregory** [comment]
> > @michael-o But enums can be extended in future releases without breaking binary compatibility. They can never be assumed to be complete and final.
> 
> Given with the current enum semantics. They are final implicitly. I cannot see in the future whether this will change. I remember my programming lecture when Sun introduced enums back then and the professor said they did not make it extendable for a good reason.
> 
> Making assumptions about likely possible features will make any design decision really hard.
> 
> > This whole enum renaming stuff is nothing more than assumptions and personal fancy.
> 
> Partially right, but can't consistency be objective in this case?

A big +1 for consistency. 

Having mentored many junior developers, I can attest that inconsistency is not good for maintenance and even guiding new code; you end up having to explain _why_ the code is inconsistent, is it by design (why?) or by lack of code review or care (lame).

> 
> > We are getting a bit side tracked here but an enum is finite by definition. Arguing they can be extended by a future release is not a valid point IMO. I can't write code today (a switch for example) for an enum value that will only exist in the future. This is obviously completely different than using a String where you can write anything you want.
> 
> +1, totally agree. Future extensions will break code. This is actually something Sun never wanted. Therefore generics have been implemented in a reified way.

**Comment by ok2c** [comment]
> Partially right, but can't consistency be objective in this case?

@michael-o I am not sure what consistency you are referring to, given that singular case convention for enums as well as their finality is just your personal opinion. 

Having said that, as long as you rename whatever stuff into some other stuff you fancy fast, I will not stand in your way. I just want this renaming party to be over. It has been going on for months already.

**Comment by michael-o** [comment]
> 
> 
> > Partially right, but can't consistency be objective in this case?
> 
> @michael-o I am not sure what consistency you are referring to, given that singular case convention for enums as well as their finality is just your personal opinion.
> 
> Having said that, as long as you rename whatever stuff into some other stuff you fancy fast, I will not stand in your way. I just want this renaming party to be over. It has been going on for months already.

Acknowledged.

**Comment by michael-o** [comment]
This will be superseded with an upcoming PR.
